### PR TITLE
permessage-deflate extension

### PR DESCRIPTION
**Pull request by adam-fowler**
Add per message deflate websocket extension
See https://www.rfc-editor.org/rfc/rfc7692 for details
- Adds a generic websocket extension protocol and websocket extension builder protocol.
- Adds permessage-deflate versions of these, using compress-nio package 

Client: Builder creates client header values, which are sent to server
Server: Builder creates server response based off client values and returns in response
Server: Builder create server version of extension
Client: Builder creates client version of extension based off what server returned

### CODE REVIEW

File: Sources/HummingbirdWSCore/PerMessageDeflateExtension.swift [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+    func shutdown() {
+        self.internalState.value.shutdown()
+    }
+
+    func processReceivedFrame(_ frame: WebSocketFrame, ws: HBWebSocket) throws -> WebSocketFrame {
+        var frame = frame
+        if frame.rsv1 {
+            let state = self.internalState.value
+            precondition(frame.fin, "Only concatenated frames with fin set can be processed by the permessage-deflate extension")
+            frame.data.writeBytes([0, 0, 255, 255])
```
Comment (Joannis) [inline Sources/HummingbirdWSCore/PerMessageDeflateExtension.swift:197]: What are these constants?

File: Sources/HummingbirdWSCore/WebSocket.swift [hunk 2]
```diff
@@ -276,8 +335,26 @@ extension HBWebSocket {
...
 }
 
-#if compiler(>=5.6)
 // HBWebSocket can be set to Sendable because ping data which is mutable is
 // managed internally and is only ever changed on the event loop
 extension HBWebSocket: @unchecked Sendable {}
-#endif // compiler(>=5.6)
+
+/// Extend WebSocketFrame to conform to CustomDebugStringConvertible
+extension WebSocketFrame: CustomDebugStringConvertible {
```
Comment (Joannis) [inline Sources/HummingbirdWSCore/WebSocket.swift:343]: While I like it, doesn't this fall under "Don't extend types you don't own with protocols you don't own"?

File: Sources/HummingbirdWSCore/PerMessageDeflateExtension.swift [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+            }
+        }
+        return frame
+    }
+
+    func processFrameToSend(_ frame: WebSocketFrame, ws: HBWebSocket) throws -> WebSocketFrame {
+        // if the frame is larger than 16 bytes, we haven't received a final frame or we are in the process of sending a message
+        // compress the data
+        let state = self.internalState.value
+        if frame.data.readableBytes > 16 || !frame.fin || state.sendState != .idle, frame.opcode == .text || frame.opcode == .binary {
```
Comment (Joannis) [inline Sources/HummingbirdWSCore/PerMessageDeflateExtension.swift:210]: Can we break this if-statement into two expressions as variables, and **then** combine them?

File: Sources/HummingbirdWSCore/PerMessageDeflateExtension.swift (hunk 1 above)
Comment (adam-fowler) [reply Sources/HummingbirdWSCore/PerMessageDeflateExtension.swift:197]: I should have added a link to the specific lines of the RFC here. It requires an empty deflate block: `00,00,ff,ff` to be added to the end of the compressed data before decompression. 

There is an equivalent operation in the compression where this block is removed.

I'll add a comment

File: Sources/HummingbirdWSCore/WebSocket.swift (hunk 2 above)
Comment (adam-fowler) [reply Sources/HummingbirdWSCore/WebSocket.swift:343]: Fair point. I think I added these when debugging the code. Maybe I should open a PR to swift-nio with a version of this

**Pull request merged**